### PR TITLE
Add flags for setting icons, sink names

### PR DESCRIPTION
**Pull request by OJFord**
For example:

    polybar-pulseaudio-control --sink-icon='⇉ ' \
      --sink-name-from='device.description' --vol-icon-low='🔈 ' \
      --vol-icon-mid='🔉 ' --vol-icon-high='🔊 ' listen

Towards #26. I haven't added options for everything since I thought it'd be better to get some early feedback, and, frankly, these are the ones that _I_ care about :smile:.

Sink nicknames from a pulseaudio property is a bit of a departure from the way it works setting within the file, but it seemed like it's be easier to use, and on my system at least has pretty usable results (to the point I almost want to suggest `device.description` as a default instead of numbers) - and perhaps there's a pulseaudio-wide way of overwriting descriptions (or other properties) that might be better than maintaining a separate list of polybar nicknames? Haven't looked into that yet.

In fact, it's surely possible, since I can change `device.description` & `bluez.alias` together for bluetooth devices by renaming them in `blueman`.

### CODE REVIEW

File: pulseaudio-control.bash
Comment (marioortizmanero) [inline pulseaudio-control.bash]: Can you add the redirection at the end for consistency?

File: pulseaudio-control.bash
Comment (marioortizmanero) [inline pulseaudio-control.bash]: These `--vol-icon-X` arguments imply that there are always going to be 3 icons for the volume. What if there was just a `--vol-icon X` that appended the icon to `VOLUME_ICONS`? Or a single `--vol-icon X,Y,Z` that split the icons at the commas. This script is meant to work for any font, and some have 2 or 4 icons for the volume.

File: pulseaudio-control.bash
Comment (marioortizmanero) [inline pulseaudio-control.bash]: I don't understand why you need `--sink-name-from`. I was expecting something as simple as `--sink-nickname "nick=id"` or similars.

Edit: Ah sorry, I've just read the explanation in your comment. It does seem to be a nice feature but we should still support custom nicknames. Unfortunately, `device.description` can't be the default for internal logic because it isn't guaranteed to be unique, and I'm not sure if *all* devices have this field. I also wouldn't want to modify any pulseaudio configuration inside this script, it's meant to be "read-only" to avoid any issues. I'd expect the user to change them outside the script, if anything.

File: pulseaudio-control.bash
Comment (marioortizmanero) [inline pulseaudio-control.bash]: I would actually prefer the options to be at the end for readability but I guess that's harder to implement, right?

File: pulseaudio-control.bash
Comment (marioortizmanero) [inline pulseaudio-control.bash]: Is it me or does this not work with arguments like `--arg value`? It only seems to work for `--arg=value`. If anything, I'd just enforce the first so that this is easier (you just match `--sink-icon` and read `$2` for the value).

**Review by marioortizmanero — CHANGES_REQUESTED** [review]

File: pulseaudio-control.bash
Comment (OJFord) [reply pulseaudio-control.bash]: Will do.

File: pulseaudio-control.bash
Comment (OJFord) [reply pulseaudio-control.bash]: Ah sure, ok.

File: pulseaudio-control.bash
Comment (OJFord) [reply pulseaudio-control.bash]: Yes I didn't mean to replace it absolutely - this was just a good enough first pass for my own use, coupled with I don't really want to hard code an ID -> nick mapping, personally. (I'd prefer to setup on a new computer, or with a new device, and still have something sane without fiddling again.)

`device.description` was just an example here, `--sink-name-from` as implemented should work with any property, and if not given (or for devices without that property) the existing nickname mapping will still be used.

> I also wouldn't want to modify any pulseaudio configuration inside this script, it's meant to be "read-only" to avoid any issues. I'd expect the user to change them outside the script, if anything.

Yes absolutely :+1: - that's what I meant - that `device.description` or even some custom `polybar.nick` could be set outside of this script, and then you run `--sink-name-from=<whatever-prop>` knowing that that's been configured elsewhere.

File: pulseaudio-control.bash
Comment (OJFord) [reply pulseaudio-control.bash]: I don't think it should be harder, just different. :slightly_smiling_face: 

In front is just my own UX preference since they affect _all_ actions, e.g. I would do:
```
foo --config-dir=fooconfig bar --maximum-bar
```
where there's flags that apply both 'globally' and just for a given subcommand 'bar'.

File: pulseaudio-control.bash
Comment (OJFord) [reply pulseaudio-control.bash]: No, it doesn't as is. I prefer `=` forms personally, and in this rudimentary parsing it makes it more robust against mis-invocations like `--arg1 --arg2 val2` (missing value to arg1).

I've refactored to support both (and it's a bit tidier anyway :slightly_smiling_face:).

File: pulseaudio-control.bash
Comment (marioortizmanero) [reply pulseaudio-control.bash]: Yes that makes sense, but not in this script specifically because we can only have a single action at a time. What I was trying to say is that `foo --a-bunch-of-options action` makes it hard to find what action is being used (which is the most important part of the command), compared to `foo action --a-bunch-of-options`. But we can keep your way just for consistency with other scripts and It's ok.

File: pulseaudio-control.bash
Comment (marioortizmanero) [reply pulseaudio-control.bash]: Your refactor looks great but it won't work when we have flags instead of arguments. For example, `--notifications` can be passed without a value, and your script will shift twice and skip the next argument.

Sorry if I'm making this too hard it's just that bash kinda sucks once the script becomes larger. At some point Python becomes a better choice haha

Comment (OJFord) [reply pulseaudio-control.bash]: :+1: good spot, sorted

File: pulseaudio-control.bash
Comment (marioortizmanero) [inline pulseaudio-control.bash]: One last thing: if the option takes a value it should be indicated clearly in the `--help` message. For example, `--vol-icons` should be `--vol-icons ICONS` or with a `=` as the delimiter. If there's no room just add a few more spaces.

**Comment by marioortizmanero** [comment]
I like your idea about `--sink-name-from`, but what if, instead of applying it with `setNickname` at the beginning of the script, we read the property as a fallback before it's displayed in `getNickname`? That way, if a new sink is added we'll get its description as well.

File: pulseaudio-control.bash
Comment (OJFord) [reply pulseaudio-control.bash]: Oops yes, thanks.

**Comment by OJFord** [comment]
Yes that's actually a problem I encountered running it today - no description when I connected my Bluetooth earphones.

I was thinking the solution would be setting `interval` to something reasonable (and putting up with that as a maximum period of no nice name) - but I prefer yours. Coming.

File: pulseaudio-control.bash
Comment (marioortizmanero) [inline pulseaudio-control.bash]: one more thing: I think this would be better by redirecting stdin into the loop instead of creating a map and reading from it:

```sh
while read -r file; do

done < <(pacmd list-sinks)
```

You could even get the property name with this:

```sh
while read -r property value; do
     case "$property" in
        name:) # ...
done < <(pacmd list-sinks)
```

File: pulseaudio-control.bash
Comment (marioortizmanero) [inline pulseaudio-control.bash]: is this echo for debugging? you might want to remove it

Comment (OJFord) [reply pulseaudio-control.bash]: No, that's the found nickname - consumed on L69.

Comment (marioortizmanero) [reply pulseaudio-control.bash]: Ah ok. Wouldn't it be better to directly return it in the `nickname` variable?

File: pulseaudio-control.bash
Comment (OJFord) [reply pulseaudio-control.bash]: Much better, thanks!

File: pulseaudio-control.bash
Comment (OJFord) [reply pulseaudio-control.bash]: I prefer to keep variables `local` to their function (and pretend I'm using a 'real' programming language :smile:) but I see that's the style used elsewhere, I'll change it for consistency.

**Comment by marioortizmanero** [comment]
Great! I think the base implementation is quite good already. If you have time, you can add the rest of the flags now (and update the README if possible) :+1:

**Comment by OJFord** [comment]
Done, and I realise now forgot the readme (sec)  - though I confess I haven't tested them all (some I'm not set up to) but they were pretty straightforward with the parsing already there.

File: README.md
Comment (OJFord) [inline README.md]: This doesn't need to be quoted, but GitHub renders it weirdly if it isn't: `--vol-max`, `--vol-icons`, and `--sink-name-from-prop` get highlighted, so it's _almost_ an accidental feature, but then `--osd` doesn't.

File: pulseaudio-control.bash
Comment (marioortizmanero) [inline pulseaudio-control.bash]: Are `--no-X` flags actually necessary if the default behavior is "turned off" by default?

File: pulseaudio-control.bash
Comment (marioortizmanero) [inline pulseaudio-control.bash]: heh I like the british touch

File: pulseaudio-control.bash
Comment (marioortizmanero) [inline pulseaudio-control.bash]: This could be `--vol-mute-color` so that it's grouped up with the rest of the related options. That, or rename `--vol-icon-mute` to `icon-vol-mute` or similars. Not sure what's better.

File: pulseaudio-control.bash
Comment (marioortizmanero) [inline pulseaudio-control.bash]: I would remove the parenthesis rather than having empty ones

File: pulseaudio-control.bash
Comment (marioortizmanero) [inline pulseaudio-control.bash]: I would remove the parenthesis here, there's no default value for `VOLUME_ICONS`.

File: pulseaudio-control.bash
Comment (marioortizmanero) [inline pulseaudio-control.bash]: Reading this after `--sink-name-from` is a bit confusing, because "name" and "nickname" are mixed. What about `--sink-nickname <id>:<nick>         nickname to assign to given sink ID (may be given multiple times)`? (or similar)

**Review by marioortizmanero — COMMENTED** [review]
Thanks! I've left some suggestions we can discuss before merging this. Sorry if I'm asking for too many changes, I want to make sure it's right :P. Also, by updating the README I meant the "Configuration", "Usage" and "Useful icons" as well. If that's too much work I can take care of it, but it should be easy.

File: pulseaudio-control.bash
Comment (OJFord) [reply pulseaudio-control.bash]: Or `(none)` perhaps? I was just aiming to show that there's nothing (well, nothing real, there is a dummy example one, hence the var not used) blacklisted by default.

File: pulseaudio-control.bash
Comment (OJFord) [reply pulseaudio-control.bash]: Well it was originally `--colour-mute` without thinking, until I got `MUTED_COLOR` from autocomplete and realised... :smile:

File: pulseaudio-control.bash
Comment (OJFord) [reply pulseaudio-control.bash]: I actually did that initially, but then I thought if unmuted colour became an option, the `--color-mute`/`--color-no-mute` (or whatever) grouping would make more sense?

File: pulseaudio-control.bash
Comment (OJFord) [reply pulseaudio-control.bash]: There is though, it's three `# `.

File: pulseaudio-control.bash
Comment (OJFord) [reply pulseaudio-control.bash]: Ah I agree, and was about to do that, but does that make it sound too much like it takes an index, or some other identifier? `pacmd` does call it a 'name'. What about changing the other instead, to `--sink-nickname-from` - since it is effectively the 'nick' of the other that it gets used as.

More I think about it that makes most sense to me - set an explicit nickname, or take it from this property.

File: pulseaudio-control.bash
Comment (OJFord) [reply pulseaudio-control.bash]: Currently no, but it allows the default to change, or for the caller to be more explicit.

**Comment by OJFord** [comment]
Oops yes, sorry I missed the 'Usage' section. Not sure what you want to change in 'Useful icons' though?

File: pulseaudio-control.bash
Comment (OJFord) [reply pulseaudio-control.bash]: Happy to remove if you prefer, but it's quite common, example from `man 1 grep`:

```
       -i, --ignore-case
              [...]

       --no-ignore-case
              [...]  This is the default. [...]
```

**Comment by marioortizmanero** [comment]
The "Useful icons" format is made for the previous configuration method. So instead of `("🕨 " "🕩 " "🕪 ")` it could be `--vol-icons="🕨 ,🕩 ,🕪 "`.

I meant something similar for the "Configuration" section too. Instead of `SINK_NICKNAME_PROP` and such it should be `--sink-nickname-prop`, because the previous one was meant for configuration by editing variables. Kinda like the "Usage" section but with more detailed explanations and in a table.

**Comment by OJFord** [comment]
Oh I see - I wasn't presuming to remove mention of the previous method (since it would still work).

Will do.

**Comment by marioortizmanero** [comment]
Yeah after this PR I don't expect anyone to configure the script by editing the top variables. Sorry if I wasn't clear with that. Having two ways to configure the script is a bit confusing IMO and using arguments is almost always better than editing the script anyway.

Yes, the Configuration section might be too repetitive now. Maybe it should be combined with the help message?

File: pulseaudio-control.bash
Comment (marioortizmanero) [inline pulseaudio-control.bash:18]: Now that I think of it, these comments at the top of the file shouldn't explain the options with comments as they aren't meant to be edited from now on. This should simply initialize the variables with the default values if anything.

File: pulseaudio-control.bash
Comment (marioortizmanero) [reply pulseaudio-control.bash]: ah ok let's just leave it like that then

File: pulseaudio-control.bash
Comment (marioortizmanero) [reply pulseaudio-control.bash]: That was just a placeholder. If anything, the `VOLUME_ICONS` default value should be empty (I'm not sure if the formatting function currently breaks if it's given an empty array).

File: pulseaudio-control.bash
Comment (marioortizmanero) [reply pulseaudio-control.bash:334]: Do we really need a `--color-no-mute` though? If anything the user can use `FFFFFF` to have white text.

File: pulseaudio-control.bash
Comment (marioortizmanero) [reply pulseaudio-control.bash]: that's ok as well, or maybe `(empty)`

File: pulseaudio-control.bash
Comment (marioortizmanero) [reply pulseaudio-control.bash]: I would make this two lines and include a comment that indicates how to obtain these nicknames with `pactl list sinks short`. That is just as useful as the previous README and clears up what the name is, IMO. We should also rename `--sink-name-from` to `--sink-nickname-from`.

File: pulseaudio-control.bash
Comment (marioortizmanero) [reply pulseaudio-control.bash]: That could just be `--default-color`. But I get your point: should we group up by the action or by what it's for? Now that I write it out, it might be better to group by the action, and rename `--vol-icons` to `--icons-vol` and similars :P You can pick anything, as long as it's consistent it's good.

**Comment by marioortizmanero** [comment]
I think this is actually ready to be merged! Thank you very much for your work :)

I will release a new version with this for those following the repo.

**Review by marioortizmanero — COMMENTED** [review]
I think this is actually ready to be merged! Thank you very much for your work :)

I will release a new version with this for those following the repo. Give me a moment to try this myself in my setup first. I'll then make a few changes updating the example module in the README, which is basically what I use.

**Comment by marioortizmanero** [comment]
Also, are you still planning on releasing this on the AUR? If so, let me know when you do. We can add that to the README, and I could help you as an Arch user with a couple packages in there as well.

**Comment by marioortizmanero** [comment]
* Should `usage` be called when an action/option is unrecognized?
* Should we have a line length limit for the options help message?

I've made some changes I wanted to do. Let me know what you think.

File: pulseaudio-control.bash
Comment (OJFord) [inline pulseaudio-control.bash]: Default should probably be the end of the third line now? Looks a little weird mid-sentence. (I realise I introduced that :blush:)

File: pulseaudio-control.bash
Comment (OJFord) [inline pulseaudio-control.bash]: I would maybe put this immediately after 'nickname to assign to given sink name', reads a bit better to me as 'nickname to assign to given sink name, taking priority over --sink-nicknames-from`.

We should probably also add the reverse, i.e. `unless overridden by --sink-nickname` to the description above too?

File: pulseaudio-control.bash
Comment (marioortizmanero) [reply pulseaudio-control.bash]: I did notice it but wasn't sure what to do. Sure, I'll add the default value at the end.

File: pulseaudio-control.bash
Comment (marioortizmanero) [reply pulseaudio-control.bash]: sure

**Comment by OJFord** [comment]
> Also, are you still planning on releasing this on the AUR?

I would like it to be available there yes, happy to package it myself or I'll of course hold off if you'd like to maintain that yourself - you'll always know when there's a new release after all :slightly_smiling_face:.

> Should usage be called when an action/option is unrecognized?

The reason I didn't was so that there's a single line of output when it's misused - reading nicely in polybar.

When I booted my PC just now for example, since I hadn't changed my main config since making these changes, polybar read 'Unrecognised option: --sink-icon'. If we printed the full usage polybar would just show the first line 'Usage: ...' which is less helpful. I suppose printing `usage` _after_ the current output would work though (not tested).

> Should we have a line length limit for the options help message?

Probably.. actually even the first one (autosync) is a little long compared to where e.g. git & grep wrap. Not sure if there's any pseudo-standard for help text specifically, but I doubt anyone would complain about 80ch.

**Comment by marioortizmanero** [comment]
> I would like it to be available there yes, happy to package it myself or I'll of course hold off if you'd like to maintain that yourself - you'll always know when there's a new release after all slightly_smiling_face.

We could just co-maintain the package, if you want (I think that's possible in the AUR).

> The reason I didn't was so that there's a single line of output when it's misused - reading nicely in polybar.
>
> When I booted my PC just now for example, since I hadn't changed my main config since making these changes, polybar read 'Unrecognised option: --sink-icon'. If we printed the full usage polybar would just show the first line 'Usage: ...' which is less helpful. I suppose printing usage after the current output would work though (not tested).

Fair enough.

> Probably.. actually even the first one (autosync) is a little long compared to where e.g. git & grep wrap. Not sure if there's any pseudo-standard for help text specifically, but I doubt anyone would complain about 80ch.

80 characters is my goto as well. It looks a bit stretched because of how long the arguments are, not sure what you think about that:

image

It uses a single echo from now on because it's much easier to edit and because of the character line at 80 chars. I've also updated the message according to your review, but I'm not sure if that's exactly what you wanted. Let me know otherwise.

I think we should also change the default value so that it's between `[` instead of `(`. The latter might be thought of as part of the explanation, while the former stands out as something different. This is just nitpicking, though.

**Comment by OJFord** [comment]
> We could just co-maintain the package, if you want (I think that's possible in the AUR).

Ah yes, so it is. :+1: 

:+1: for single echo, I was tempted to do that too.

Screenshot looks ok to me, maybe gets a little hard to read at the bottom over several lines, but then spaces around that and not the others might look weirder. (Also nitpicking though!) Think I agree about `[]` defaults.

**Comment by marioortizmanero** [comment]
I've ended up using an indentation of 2 spaces instead of 4 for the help message, which gives *just a bit* of space more (not sure what else can be done). I'll merge this and we can discuss the PKGBUILD at #35.

Thanks a lot for your help with this PR!

**Comment by OJFord** [comment]
Awesome, thanks!

**Pull request merged**
